src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -691,27 +712,33 @@ def __copy__(self):
 
         return new
 
-    def __deepcopy__(self, memodict={}) -> "SelectStatement":  # noqa: B006
+    def __deepcopy__(self, memodict=None) -> "SelectStatement":  # noqa: B006
+        if memodict is None:
+            memodict = {}
+        if (self_id := id(self)) in memodict:
+            return memodict[self_id]
         copied = SelectStatement(
-            projection=deepcopy(self.projection),
-            from_=deepcopy(self.from_),
-            where=deepcopy(self.where),
-            order_by=deepcopy(self.order_by),
-            limit_=deepcopy(self.limit_),
+            projection=deepcopy(self.projection, memodict),
```

Comment:
another simple optimizaiton we can do is that expression doesn't really need a copy, because we are not really doing any update to expressions, but this could be little bit wired, i think we can keep the deep copy here for now

Reply:
`ScalarSubQueryExpression` contains a plan. keeping deepcopy here is better.